ce/ce/artifacts/activation.ts

```diff
@@ -77,90 +571,197 @@ export class Activation {
     return toXml(msbuildFile);
   }
 
-  /** a collection of #define declarations that would assumably be applied to all compiler calls. */
-  defines = new Map<string, string>();
+  protected async generateEnvironmentVariables(originalEnvironment: Record<string, string | undefined>): Promise<[Record<string, string>, Record<string, string>]> {
 
-  /** a collection of tool definitions from artifacts (think shell 'aliases')  */
-  tools = new Map<string, string>();
+    const undo = new Record<string, string>();
+    const env = new Record<string, string>();
 
-  /** Aliases are tools that get exposed to the user as shell aliases */
-  aliases = new Map<string, string>();
+    for await (const [pathVariable, locations] of this.paths) {
+      if (locations.size) {
+        const originalVariable = linq.find(originalEnvironment, pathVariable) || '';
+        if (originalVariable) {
+          for (const p of originalVariable.split(delimiter)) {
+            if (p) {
+              locations.add(p);
+            }
+          }
+        }
+        // compose the final value
+        env[pathVariable] = linq.join(locations, delimiter);
 
-  /** a collection of 'published locations' from artifacts. useful for msbuild */
-  locations = new Map<string, Uri>();
+        // set the undo data
+        undo[pathVariable] = originalVariable || '';
+      }
+    }
 
-  /** a collection of environment variables from artifacts that are intended to be combinined into variables that have PATH delimiters */
-  paths = new Map<string, Array<Uri>>();
+    // combine environment variables with multiple values with spaces (uses: CFLAGS, etc)
+    for await (const [variable, values] of this.environmentVariables) {
+      env[variable] = linq.join(values, ' ');
+      undo[variable] = originalEnvironment[variable] || '';
+    }
 
-  /** environment variables from artifacts */
-  environment = new Map<string, Array<string>>();
+    // .tools get defined as environment variables too.
+    for await (const [variable, value] of this.tools) {
+      env[variable] = value;
+      undo[variable] = originalEnvironment[variable] || '';
+    }
 
-  /** a collection of arbitrary properties from artifacts. useful for msbuild */
-  properties = new Map<string, Array<string>>();
+    // .defines get compiled into a single environment variable.
+    if (this.definesCount) {
+      let defines = '';
+      for await (const [name, value] of this.defines) {
+        defines += value ? `-D ${name}=${value} ` : `-D ${name} `;
```

Comment:
```suggestion
        defines += value ? `-D${name}=${value} ` : `-D${name} `;
```

Traditionally there isn't supposed to be a space here although I suspect most compilers don't care.

Reply:
I will defer to you on that - I don't know for certain.